mtags/src/main/scala-3/scala/meta/internal/pc/SelectionRangeProvider.scala

```diff
@@ -85,3 +99,33 @@ class SelectionRangeProvider(
       child
 
 end SelectionRangeProvider
+
+object SelectionRangeProvider:
+  import scala.meta.*
+  import scala.meta.Token.Comment
+
+  /** check pos.start,if it is within comment then expand to comment */
+  def getCommentRanges(param: OffsetParams, start: Int, end: Int) =
+    // overwrite implicit 'implicit def current: Scala213' in scalameta
+    implicit val dialect: Dialect = dialects.Scala3.withAllowToplevelTerms(true)
+    val programStr = param.text()
+    // deal with multiple source text. probably a better way ?
+    val tkSrc = programStr.parse[Source].toOption.map(_.tokens)
+    val tkExpr = programStr.parse[Stat].toOption.map(_.tokens)
+    val tkType = programStr.parse[Type].toOption.map(_.tokens)
+
+    val tokens = tkSrc orElse tkExpr orElse tkType // orElse tkTopDecls
```

Comment:
You should use something along the lines of:
```
val (treeStart, treeEnd) = path.headOption
      .map(t => (t.pos.start, t.pos.end))
      .getOrElse((0, text.size))
val offset = pos.start
val tokens = text.slice(treeStart, treeEnd)tokenize.toOption
      tokens
        .flatMap(t =>
          t.collectFirst {
            case t: Token.Comment
                if treeStart + t.pos.start < currentOffset &&
                  treeStart + t.pos.end >= currentOffset =>
              val range = new SelectionRange()
              range.setRange(t.pos.toLSP)
              Some(range)
          
            case _ =>
              Noene
          }
        )
```

Reply:
Thanks for suggestion,but i'm not sure how to update offset to **input** ,which will cause error if it's not updated  (used later in `def toLsp: l.Range` of `extension (pos: SourcePosition)`):
```scala
(commentList) // ++ merged)
        .filter((commentStart, commentEnd, _) =>
          commentStart <= cursorStartShifted && cursorStartShifted <= commentEnd
        )
        .map { (commentStart, commentEnd, oldPos) =>
          // todo! need to upg input if offset is nonzero!
          val newPos: Position =
            meta.Position.Range(
              oldPos.input,
              commentStart + offsetStart,
              commentEnd + offsetStart,
            )
          newPos
        }
```